### ISSUE TITLE
Why Pull operations is so Slow?

### ISSUE THREAD

**Issue by Stellacci** [libgit2/libgit2sharp#1527]
Hi all, I'm new and I'm developing using this library. I have some problem with pull operations. This is my code:

```
 Debug.Write("Pulling branch. . .");
            LibGit2Sharp.PullOptions options = new LibGit2Sharp.PullOptions();
            options.FetchOptions = new FetchOptions();
            options.FetchOptions.CredentialsProvider = new CredentialsHandler(
                (url, usernameFromUrl, types) =>
                    new UsernamePasswordCredentials()
                    {
                        Username = this.Username,
                        Password = this.Password
                    });
            Signature author = new Signature(this.Username, this.EmailUser, DateTime.Now);
            Commands.Pull(Repo, author, options);
            Debug.Write("Pulling done.");
```
I don't uderstand why Pulling ONE FILE (txt 1kb) is so SLOW. Is there anyone help me?

**Comment by ethomson** [comment]
You've given literally no data to work with that might answer your question.  What do you consider "slow"?  How fast does git work?  What version of git are you comparing against?  Is this a large repository?  Are you transferring the same data with both clients?  Are you sure?  You don't have any callbacks to show how many objects you're fetching...

If you can provide a sample repository that illustrates that this is fast in one scenario and slow in another, that would be helpful.  Otherwise, there is no way to answer your question.

**Comment by Stellacci** [comment]
If I add a simple .txt file in the remote repository, I expect that the pulling of ONLY THAT file is fast. Instead it takes about 10-15 to finish the pull.

**Comment by ethomson** [comment]
Yes, I understand how pull works.  You haven't answered any of my questions.

**Comment by Stellacci** [comment]
Ok, sorry.
-I'm using libgit2sharp version 0.24
-The repository is very large
-I'm not trasnferring any data

**Comment by ethomson** [comment]
> The repository is very large

How large are we talking?  How many branches and tags do you have?

How long does git take to do the same pull?

**Comment by ethomson** [comment]
To clarify: the problem isn't _how much data_ is being transferred, it's _walking the graph_ to figure out _what_ needs to be transferred.

**Comment by Stellacci** [comment]
- The repository is large about 10 GB
- GIT takes 3 seconds to do the same pull

**Comment by ethomson** [comment]
OK, and it takes LibGit2Sharp to take 10-15 _seconds_ or _minutes_?  (There's no unit in your comment above.)

**Comment by Stellacci** [comment]
Minutes (I forget to write it)

**Comment by Stellacci** [comment]
Do you have any idea?

**Comment by ethomson** [comment]
Not offhand, no.  I suspect that libgit2 is walking much more of the graph than is needed, but I don't know why it would be doing that.

**Comment by carlosmn** [comment]
Are you doing this over the network? How are you running `git`? Are they equivalent calls?

**Comment by Stellacci** [comment]
@carlosmn Yes, I do over the network. But I think that my code is wrong maybe. Is it correct or I can add some fixes?

**Comment by Stellacci** [comment]
Anything else?

**Comment by Stellacci** [comment]
@carlosmn  Do you have some idea about my problem?

**Comment by ethomson** [comment]
> How are you running `git`?

**Comment by Stellacci** [comment]
With "GitExtension" tool. But this does not explain why the pull in Libgit2sharp is slow
